framework/src/base/MeshGeneratorSystem.C

```diff
@@ -0,0 +1,496 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MeshGeneratorSystem.h"
+
+#include "MooseApp.h"
+#include "MeshGenerator.h"
+#include "DependencyResolver.h"
+
+MeshGeneratorSystem::MeshGeneratorSystem(MooseApp & app)
+  : PerfGraphInterface(app.perfGraph(), "MeshGeneratorSystem"),
+    ParallelObject(app),
+    _app(app),
+    _popped_final_mesh_generator(false)
+{
+}
+
+void
+MeshGeneratorSystem::addMeshGenerator(const std::string & type,
+                                      const std::string & name,
+                                      const InputParameters & params)
+{
+  mooseAssert(!_mesh_generator_params.count(name), "Already exists");
+  _mesh_generator_params.emplace(
+      std::piecewise_construct, std::forward_as_tuple(name), std::forward_as_tuple(type, params));
+
+  // This should be a sub mesh generator. We can assume this because if we're in the middle of
+  // constructing mesh generators (not "adding" them, where we simply store their parameters),
+  // that we're constructing a sub generator also in the correct order
+  if (_app.constructingMeshGenerators())
+    createMeshGenerator(name);
+}
+
+const MeshGenerator &
+MeshGeneratorSystem::appendMeshGenerator(const std::string & type,
+                                         const std::string & name,
+                                         InputParameters params)
+{
+  if (!appendingMeshGenerators())
+    mooseError("Can only call appendMeshGenerator() during the append_mesh_generator task");
+
+  // Make sure this mesh generator has one and _only_ one input, as "input"
+  const auto param_name_mg_name_pairs = getMeshGeneratorParamDependencies(params, true);
+  if (param_name_mg_name_pairs.size() != 1 || param_name_mg_name_pairs[0].first != "input")
+    mooseError("While adding ",
+               type,
+               " '",
+               name,
+               "' via appendMeshGenerator():\nCan only append a mesh generator that takes a "
+               "single input mesh generator via the parameter named 'input'");
+
+  // If no final generator is set, we need to make sure that we have one; we will hit
+  // this the first time we add an appended MeshGenerator and only need to do it once.
+  // We'll then generate the final ordering within the execution phase. We'll also
+  // clear the ordering because it isn't particularly valid anymore once we add this one.
```

Comment:
why is it not valid?

Reply:
is the final one not properly at the end?

Reply:
We could either:

- call order every time we append a new generator, and then add special logic to check if we need to call it in execute
- call order when we append the first generator, keep adding on the final name as we append more, and then call it once in execute without any fancy logic

The second is better because the logic is more straight forward, and clearing during the append phase makes it clear and reminds us that we need to re-order during execute

Reply:
ok as it is it works better to not re-order in case the user is appending multiple times. 
good that way